## Segmenting and Clustering Neighborhoods in Toronto

##### We need to import in the associated libraries for scraping web data (urlib) and reading the html data (Beautiful Soup)

In [48]:
import urllib.request
from bs4 import BeautifulSoup


##### Define the url and put it into a variable

In [49]:
url='https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
url

'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'

##### Open the defined url variable and put it into a variable using the urlib request

In [50]:
html_page = urllib.request.urlopen(url)
html_page

##### Next we need to parse the html using beautiful soup and put it into a variable

In [51]:
html_postal_codes = BeautifulSoup(html_page, "lxml")


##### Find the table in the html code after exploring the html

In [52]:
postal_table=html_postal_codes.find('table', class_='wikitable sortable')


##### Notice above we have three identifiers in the postal html, Postal Code, Borough, and Neighborhood identified in the header as th which repeat every 3 lines for the detail td..  Create empty lists and run a loop through the html code matching up every 3 lines and append the data to the lists.

In [53]:
#create empty lists for each segment
postal_code=[]
borough=[]
neighborhood=[]

#run a loop through to append data to the lists from the html for each td element
for row in postal_table.findAll('tr'):
    cells=row.findAll('td')
    if len(cells)==3:
        postal_code.append(cells[0].find(text=True))
        borough.append(cells[1].find(text=True))
        neighborhood.append(cells[2].find(text=True))


##### Next after reviewing we see a lot of new lines which need to be stripped out from the data, strip these out for each list
       

In [54]:
#Strip out the new lines from the list
postal_code = [ p.strip() for p in postal_code ]
borough = [ b.strip() for b in borough ]
neighborhood = [n.strip() for n in neighborhood]


#### Create a pandas data frame and map the data to the data frame

In [55]:
#Assign the data to a pandas data frameand map the data to columns in this data frame
import pandas as pd
df=pd.DataFrame(postal_code,columns=['PostalCode'])
df['Borough']=borough
df['Neighborhood']=neighborhood
#Convert the slashes to coomas in Neighborhood
df['Neighborhood']=df['Neighborhood'].str.replace('/',',')

#Filter out the Not Assigned from the Borough
df=df[df.Borough != 'Not assigned']

df.head(30)


PostalCode           Borough  \
2         M3A        North York   
3         M4A        North York   
4         M5A  Downtown Toronto   
5         M6A        North York   
6         M7A  Downtown Toronto   
8         M9A         Etobicoke   
9         M1B       Scarborough   
11        M3B        North York   
12        M4B         East York   
13        M5B  Downtown Toronto   
14        M6B        North York   
17        M9B         Etobicoke   
18        M1C       Scarborough   
20        M3C        North York   
21        M4C         East York   
22        M5C  Downtown Toronto   
23        M6C              York   
26        M9C         Etobicoke   
27        M1E       Scarborough   
30        M4E      East Toronto   
31        M5E  Downtown Toronto   
32        M6E              York   
36        M1G       Scarborough   
39        M4G         East York   
40        M5G  Downtown Toronto   
41        M6G  Downtown Toronto   
45        M1H       Scarborough   
46        M2H        North York   
47        M3H        North York   
48        M4H         East York   

                                         Neighborhood  
2                                           Parkwoods  
3                                    Victoria Village  
4                          Regent Park , Harbourfront  
5                   Lawrence Manor , Lawrence Heights  
6        Queen's Park , Ontario Provincial Government  
8                                    Islington Avenue  
9                                     Malvern , Rouge  
11                                          Don Mills  
12                   Parkview Hill , Woodbine Gardens  
13                           Garden District, Ryerson  
14                                          Glencairn  
17  West Deane Park , Princess Gardens , Martin Gr...  
18           Rouge Hill , Port Union , Highland Creek  
20                                          Don Mills  
21                                   Woodbine Heights  
22                                     St. James Town  
23                                 Humewood-Cedarvale  
26  Eringate , Bloordale Gardens , Old Burnhamthor...  
27                Guildwood , Morningside , West Hill  
30                                        The Beaches  
31                                        Berczy Park  
32                                Caledonia-Fairbanks  
36                                             Woburn  
39                                            Leaside  
40                                 Central Bay Street  
41                                           Christie  
45                                          Cedarbrae  
46                                  Hillcrest Village  
47  Bathurst Manor , Wilson Heights , Downsview North  
48                                   Thorncliffe Park

In [56]:
df.shape

(103, 3)

## Second Part of the Project find the Lattiitude and Longitude for each postal code

#### We now need to read in the csv file that has latitude and longitude coordinates by postal code into a pandas data frame so that we can begin to combine it with our existing dataset. I also needed to rename the Postal Code column so column names match between data frames

In [57]:
lat_long=pd.read_csv('Geospatial_Coordinates.csv')
lat_long.rename(columns = {'Postal Code':'PostalCode'}, inplace=True)
lat_long.head()

PostalCode   Latitude  Longitude
0        M1B  43.806686 -79.194353
1        M1C  43.784535 -79.160497
2        M1E  43.763573 -79.188711
3        M1G  43.770992 -79.216917
4        M1H  43.773136 -79.239476

#### Now that we have the data in a dataframe we can merge it with our current dataframe we will do an inner join based on Postal Code

In [58]:
df2=pd.merge(df,lat_long, on='PostalCode', how='inner')

In [59]:
df2.head(10)

PostalCode           Borough                                  Neighborhood  \
0        M3A        North York                                     Parkwoods   
1        M4A        North York                              Victoria Village   
2        M5A  Downtown Toronto                    Regent Park , Harbourfront   
3        M6A        North York             Lawrence Manor , Lawrence Heights   
4        M7A  Downtown Toronto  Queen's Park , Ontario Provincial Government   
5        M9A         Etobicoke                              Islington Avenue   
6        M1B       Scarborough                               Malvern , Rouge   
7        M3B        North York                                     Don Mills   
8        M4B         East York              Parkview Hill , Woodbine Gardens   
9        M5B  Downtown Toronto                      Garden District, Ryerson   

    Latitude  Longitude  
0  43.753259 -79.329656  
1  43.725882 -79.315572  
2  43.654260 -79.360636  
3  43.718518 -79.464763  
4  43.662301 -79.389494  
5  43.667856 -79.532242  
6  43.806686 -79.194353  
7  43.745906 -79.352188  
8  43.706397 -79.309937  
9  43.657162 -79.378937

In [13]:
df2.shape

(103, 5)

## Third Part of the Project - Cluster Neighborhoods and Analyze

#### We will need some new libraries for this part of the project

In [14]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')

Solving environment: done

# All requested packages already installed.

Solving environment: done

# All requested packages already installed.

Libraries imported.


##### Lets take a look at the unique boroughs to determine which boroughs we will decide to explore

In [60]:
df2.Borough.unique()

array(['North York', 'Downtown Toronto', 'Etobicoke', 'Scarborough',
       'East York', 'York', 'East Toronto', 'West Toronto',
       'Central Toronto', 'Mississauga'], dtype=object)

#### We will start with just a segment of our data, lets start with the bouroughs of Toronto (Downtown Toronto, East Toronto, Central Toronto, and West Toronto) which we will call Toronto. We will create another data fram with these bouroughs

In [61]:
toronto=['Downtown Toronto','East Toronto','West Toronto','Central Toronto']
toronto=df2[df2['Borough'].isin(toronto)].reset_index(drop=True)
toronto.head()

PostalCode           Borough                                  Neighborhood  \
0        M5A  Downtown Toronto                    Regent Park , Harbourfront   
1        M7A  Downtown Toronto  Queen's Park , Ontario Provincial Government   
2        M5B  Downtown Toronto                      Garden District, Ryerson   
3        M5C  Downtown Toronto                                St. James Town   
4        M4E      East Toronto                                   The Beaches   

    Latitude  Longitude  
0  43.654260 -79.360636  
1  43.662301 -79.389494  
2  43.657162 -79.378937  
3  43.651494 -79.375418  
4  43.676357 -79.293031

#### Lets look at the shape of the data

In [63]:
toronto.shape

(39, 5)

#### As we can see from above we will have 39 neighborhoods in our data. Lets now set this data up for plotting. We will pick up a certain point for to gather the latitude and longitude points

In [64]:
address = 'Downtown Toronto, Garden District, ON'

geolocator = Nominatim(user_agent="ca_explorer")
location = geolocator.geocode(address)
latitude_ca = location.latitude
longitude_ca = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude_ca, longitude_ca))

The geograpical coordinate of Toronto are 43.6531818, -79.3751072.


In [65]:
# create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[latitude_ca, longitude_ca], zoom_start=12)

# add markers to map
for lat, lng, borough, neighborhood in zip(toronto['Latitude'], toronto['Longitude'], toronto['Borough'], downtown_toronto['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

#### Now lets set up our connection to the foursquare api so that we can explore nearby venue locations within a certain distance of our neighborhoods.

In [20]:
CLIENT_ID = 'WYN1IAIDVPDMG2L1TYL0XUCQ3ENZGPLOTF2K4J05GGP44JIX' # your Foursquare ID
CLIENT_SECRET = 'KTOBN45FG3TIL5ZPB3TLP0TINCHFK2Q3ZH3J5X133K5BZ4RD' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: WYN1IAIDVPDMG2L1TYL0XUCQ3ENZGPLOTF2K4J05GGP44JIX
CLIENT_SECRET:KTOBN45FG3TIL5ZPB3TLP0TINCHFK2Q3ZH3J5X133K5BZ4RD


#### Lets identify a point within our data set and run through the foursquare api to get a sample of data to review

In [69]:
toronto.loc[0,'Neighborhood']

'Regent Park , Harbourfront'

In [70]:
neighborhood_latitude = toronto.loc[0, 'Latitude'] # neighborhood latitude value
neighborhood_longitude = toronto.loc[0, 'Longitude'] # neighborhood longitude value

neighborhood_name = toronto.loc[0, 'Neighborhood'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_latitude, 
                                                               neighborhood_longitude))

Latitude and longitude values of Regent Park , Harbourfront are 43.6542599, -79.3606359.


In [71]:
LIMIT=100
radius=500
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(CLIENT_ID,CLIENT_SECRET,VERSION,neighborhood_latitude,neighborhood_longitude,radius,LIMIT)
url



'https://api.foursquare.com/v2/venues/explore?&client_id=WYN1IAIDVPDMG2L1TYL0XUCQ3ENZGPLOTF2K4J05GGP44JIX&client_secret=KTOBN45FG3TIL5ZPB3TLP0TINCHFK2Q3ZH3J5X133K5BZ4RD&v=20180605&ll=43.6542599,-79.3606359&radius=500&limit=100'

### Lets review the sample json response from our call

In [72]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5ea4759f660a9f001b2288bd'},
 'response': {'suggestedFilters': {'header': 'Tap to show:',
   'filters': [{'name': 'Open now', 'key': 'openNow'}]},
  'headerLocation': 'Corktown',
  'headerFullLocation': 'Corktown, Toronto',
  'headerLocationGranularity': 'neighborhood',
  'totalResults': 48,
  'suggestedBounds': {'ne': {'lat': 43.6587599045, 'lng': -79.3544279001486},
   'sw': {'lat': 43.6497598955, 'lng': -79.36684389985142}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '54ea41ad498e9a11e9e13308',
       'name': 'Roselle Desserts',
       'location': {'address': '362 King St E',
        'crossStreet': 'Trinity St',
        'lat': 43.653446723052674,
        'lng': -79.3620167174383,
        'labeledLatLngs': [{'label': 'display',
 

#### lets create a function to captures the categories from the response

In [25]:
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

#### Now lets flatten the json response and apply the get category type function to the data

In [73]:
venues = results['response']['groups'][0]['items']
   
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']

nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head(50)



name             categories        lat  \
0                       Roselle Desserts                 Bakery  43.653447   
1                          Tandem Coffee            Coffee Shop  43.653559   
2                     Morning Glory Cafe         Breakfast Spot  43.653947   
3                 Cooper Koo Family YMCA    Distribution Center  43.653249   
4                    Body Blitz Spa East                    Spa  43.654735   
5                         Impact Kitchen             Restaurant  43.656369   
6                        Corktown Common                   Park  43.655618   
7                     The Extension Room   Gym / Fitness Center  43.653313   
8       The Distillery Historic District          Historic Site  43.650244   
9                 Figs Breakfast & Lunch         Breakfast Spot  43.655675   
10              Distillery Sunday Market         Farmers Market  43.650075   
11                   SOMA chocolatemaker         Chocolate Shop  43.650622   
12                        Rooster Coffee            Coffee Shop  43.651900   
13                             Starbucks            Coffee Shop  43.651613   
14              Dominion Pub and Kitchen                    Pub  43.656919   
15                                  Arvo            Coffee Shop  43.649963   
16  Young Centre for the Performing Arts  Performing Arts Venue  43.650825   
17                        Underpass Park                   Park  43.655764   
18                              Cacao 70           Dessert Shop  43.650067   
19                      Cocina Economica     Mexican Restaurant  43.654959   
20                       Sumach Espresso            Coffee Shop  43.658135   
21                   Brick Street Bakery                 Bakery  43.650574   
22            Cluny Bistro & Boulangerie      French Restaurant  43.650565   
23               Dark Horse Espresso Bar            Coffee Shop  43.653081   
24                Parliament Square Park                   Park  43.650264   
25                       The Yoga Lounge            Yoga Studio  43.655515   
26                           Caffe Furbo                   Café  43.649970   
27                       Alumnae Theatre                Theater  43.652756   
28                       Berkeley Church            Event Space  43.655123   
29                      Greg's Ice Cream         Ice Cream Shop  43.650102   
30                    Soulpepper Theatre                Theater  43.650780   
31                    John Fluevog Shoes             Shoe Store  43.649896   
32                       ODIN Cafe + Bar                   Café  43.656739   
33           The Sweet Escape Patisserie                 Bakery  43.650632   
34                     Mill St. Brew Pub                    Pub  43.650353   
35                            The Aviary                    Pub  43.653634   
36                          Arta Gallery            Art Gallery  43.650022   
37                                 Izumi       Asian Restaurant  43.649970   
38           The Abnormal Beauty Company         Cosmetics Shop  43.649892   
39                       TD Canada Trust                   Bank  43.652780   
40                                Vistek      Electronics Store  43.657046   
41                       Savoury Grounds            Coffee Shop  43.656821   
42                        The Beer Store             Beer Store  43.657773   
43                           Flame Shack             Restaurant  43.656844   
44         Residence & Conference Centre                  Hotel  43.653040   
45                      The Healthy Road      Health Food Store  43.656265   
46                             Wine Rack              Wine Shop  43.656573   
47                            GW General           Antique Shop  43.650495   

          lng  
0  -79.362017  
1  -79.361809  
2  -79.361149  
3  -79.358008  
4  -79.359874  
5  -79.356980  
6  -79.356211  
7  -79.359725  
8  -79.359323  
9  -79.364503  
10 -79.361832  
11 -79.358127  
12 -7

In [74]:
print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))

48 venues were returned by Foursquare.


#### Now lets define a function that will run against the foursquare api and will be able to loop through our neighborhoods and append nearby locations

In [75]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

#### Lets call the getnearbyvenues function and pass in our neighborhoods and longitudes

In [76]:
downtown_venues = getNearbyVenues(names=toronto['Neighborhood'],
                                   latitudes=toronto['Latitude'],
                                   longitudes=toronto['Longitude']
                                  )

Regent Park , Harbourfront
Queen's Park , Ontario Provincial Government
Garden District, Ryerson
St. James Town
The Beaches
Berczy Park
Central Bay Street
Christie
Richmond , Adelaide , King
Dufferin , Dovercourt Village
Harbourfront East , Union Station , Toronto Islands
Little Portugal , Trinity
The Danforth West , Riverdale
Toronto Dominion Centre , Design Exchange
Brockton , Parkdale Village , Exhibition Place
India Bazaar , The Beaches West
Commerce Court , Victoria Hotel
Studio District
Lawrence Park
Roselawn
Davisville North
Forest Hill North & West
High Park , The Junction South
North Toronto West
The Annex , North Midtown , Yorkville
Parkdale , Roncesvalles
Davisville
University of Toronto , Harbord
Runnymede , Swansea
Moore Park , Summerhill East
Kensington Market , Chinatown , Grange Park
Summerhill West , Rathnelly , South Hill , Forest Hill SE , Deer Park
CN Tower , King and Spadina , Railway Lands , Harbourfront West , Bathurst Quay , South Niagara , Island airport
Roseda

#### Lets look at the shape of our data and I like coffee so let me see how many coffee shops are out there

In [77]:
print('Total Shape :', downtown_venues.shape)


coffee_shops=downtown_venues[downtown_venues['Venue Category']=='Coffee Shop']
print('Coffee Shops Shape :', coffee_shops.shape)

Total Shape : (1612, 7)
Coffee Shops Shape : (143, 7)


#### Just curious which neighborhood can I make sure I can find coffee

In [78]:
coffee_shops.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Berczy Park                                                             4   
Brockton , Parkdale Village , Exhibition Place                          2   
CN Tower , King and Spadina , Railway Lands , H...                      1   
Central Bay Street                                                     13   
Christie                                                                1   
Church and Wellesley                                                    5   
Commerce Court , Victoria Hotel                                        11   
Davisville                                                              2   
First Canadian Place , Underground city                                10   
Garden District, Ryerson                                                8   
Harbourfront East , Union Station , Toronto Isl...                     13   
Kensington Market , Chinatown , Grange Park                             4   
Little Portugal , Trinity                                               1   
North Toronto West                                                      2   
Parkdale , Roncesvalles                                                 1   
Queen's Park , Ontario Provincial Government                           10   
Regent Park , Harbourfront                                              7   
Richmond , Adelaide , King                                              8   
Runnymede , Swansea                                                     3   
St. James Town                                                          5   
St. James Town , Cabbagetown                                            3   
Stn A PO Boxes                                                          8   
Studio District                                                         3   
Summerhill West , Rathnelly , South Hill , Fore...                      2   
The Annex , North Midtown , Yorkville                                   2   
The Danforth West , Riverdale                                           3   
Toronto Dominion Centre , Design Exchange                              11   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Berczy Park                                                              4   
Brockton , Parkdale Village , Exhibition Place                           2   
CN Tower , King and Spadina , Railway Lands , H...                       1   
Central Bay Street                                                      13   
Christie                                                                 1   
Church and Wellesley                                                     5   
Commerce Court , Victoria Hotel                                         11   
Davisville                                                               2   
First Canadian Place , Underground city                                 10   
Garden District, Ryerson                                                 8   
Harbourfront East , Union Station , Toronto Isl...                      13   
Kensington Market , Chinatown , Grange Park                              4   
Little Portugal , Trinity                                                1   
North Toronto West                                                       2   
Parkdale , Roncesvalles                                                  1   
Queen's Park , Ontario Provincial Government                            10   
Regent Park , Harbourfront                                               7   
Richmond , Adelaide , King                                               8   
Runnymede , Swansea                                                      3   
St. James Town                                                           5   
St. James Town , Cabbagetown                                             3   
Stn A PO Boxes          

In [79]:
print('There are {} uniques categories.'.format(len(downtown_venues['Venue Category'].unique())))

There are 231 uniques categories.


#### Now lets convert the categories to columns using one hot encoding

In [80]:
# one hot encoding
downtown_onehot = pd.get_dummies(downtown_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
downtown_onehot['Neighborhoods'] = downtown_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [downtown_onehot.columns[-1]] + list(downtown_onehot.columns[:-1])
downtown_onehot = downtown_onehot[fixed_columns]
downtown_onehot.drop(['Neighborhood'],axis=1,inplace=True)
downtown_onehot.head(30)

Neighborhoods  Airport  Airport Food Court  Airport Gate  \
0   Regent Park , Harbourfront        0                   0             0   
1   Regent Park , Harbourfront        0                   0             0   
2   Regent Park , Harbourfront        0                   0             0   
3   Regent Park , Harbourfront        0                   0             0   
4   Regent Park , Harbourfront        0                   0             0   
5   Regent Park , Harbourfront        0                   0             0   
6   Regent Park , Harbourfront        0                   0             0   
7   Regent Park , Harbourfront        0                   0             0   
8   Regent Park , Harbourfront        0                   0             0   
9   Regent Park , Harbourfront        0                   0             0   
10  Regent Park , Harbourfront        0                   0             0   
11  Regent Park , Harbourfront        0                   0             0   
12  Regent Park , Harbourfront        0                   0             0   
13  Regent Park , Harbourfront        0                   0             0   
14  Regent Park , Harbourfront        0                   0             0   
15  Regent Park , Harbourfront        0                   0             0   
16  Regent Park , Harbourfront        0                   0             0   
17  Regent Park , Harbourfront        0                   0             0   
18  Regent Park , Harbourfront        0                   0             0   
19  Regent Park , Harbourfront        0                   0             0   
20  Regent Park , Harbourfront        0                   0             0   
21  Regent Park , Harbourfront        0                   0             0   
22  Regent Park , Harbourfront        0                   0             0   
23  Regent Park , Harbourfront        0                   0             0   
24  Regent Park , Harbourfront        0                   0             0   
25  Regent Park , Harbourfront        0                   0             0   
26  Regent Park , Harbourfront        0                   0             0   
27  Regent Park , Harbourfront        0                   0             0   
28  Regent Park , Harbourfront        0                   0             0   
29  Regent Park , Harbourfront        0                   0             0   

    Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0                0                0                 0                    0   
1                0                0                 0                    0   
2                0                0                 0                    0   
3                0                0                 0                    0   
4                0                0                 0                    0   
5                0                0                 0                    0   
6                0                0                 0                    0   
7                0                0                 0                    0   
8                0                0                 0                    0   
9                0                0                 0                    0   
10               0                0                 0                    0   
11               0                0                 0                    0   
12               0                0                 0                    0   
13               0                0                 0                    0   
14               0                0                 0                    0   
15               0                0                 0                    0   
16               0                0                 0                    0   
17               0                0                 0                    0   
18               0                0                 0                    0   
19               0                0                 0                

In [81]:
downtown_onehot.shape

(1612, 231)

#### Lets group the neighborhoods and find the mean of each category by neighborhood

In [82]:
downtown_grouped = downtown_onehot.groupby('Neighborhoods').mean().reset_index()
downtown_grouped

Neighborhoods   Airport  \
0                                         Berczy Park  0.000000   
1      Brockton , Parkdale Village , Exhibition Place  0.000000   
2               Business reply mail Processing CentrE  0.000000   
3   CN Tower , King and Spadina , Railway Lands , ...  0.055556   
4                                  Central Bay Street  0.000000   
5                                            Christie  0.000000   
6                                Church and Wellesley  0.000000   
7                     Commerce Court , Victoria Hotel  0.000000   
8                                          Davisville  0.000000   
9                                    Davisville North  0.000000   
10                      Dufferin , Dovercourt Village  0.000000   
11            First Canadian Place , Underground city  0.000000   
12                           Forest Hill North & West  0.000000   
13                           Garden District, Ryerson  0.000000   
14  Harbourfront East , Union Station , Toronto Is...  0.000000   
15                     High Park , The Junction South  0.000000   
16                    India Bazaar , The Beaches West  0.000000   
17        Kensington Market , Chinatown , Grange Park  0.000000   
18                                      Lawrence Park  0.000000   
19                          Little Portugal , Trinity  0.000000   
20                       Moore Park , Summerhill East  0.000000   
21                                 North Toronto West  0.000000   
22                            Parkdale , Roncesvalles  0.000000   
23       Queen's Park , Ontario Provincial Government  0.000000   
24                         Regent Park , Harbourfront  0.000000   
25                         Richmond , Adelaide , King  0.000000   
26                                           Rosedale  0.000000   
27                                           Roselawn  0.000000   
28                                Runnymede , Swansea  0.000000   
29                                     St. James Town  0.000000   
30                       St. James Town , Cabbagetown  0.000000   
31                                     Stn A PO Boxes  0.000000   
32                                    Studio District  0.000000   
33  Summerhill West , Rathnelly , South Hill , For...  0.000000   
34              The Annex , North Midtown , Yorkville  0.000000   
35                                        The Beaches  0.000000   
36                      The Danforth West , Riverdale  0.000000   
37          Toronto Dominion Centre , Design Exchange  0.000000   
38                    University of Toronto , Harbord  0.000000   

    Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0             0.000000      0.000000        0.000000         0.000000   
1             0.000000      0.000000        0.000000         0.000000   
2             0.000000      0.000000        0.000000         0.000000   
3             0.055556      0.055556        0.111111         0.166667   
4             0.000000      0.000000        0.000000         0.000000   
5             0.000000      0.000000        0.000000         0.000000   
6             0.000000      0.000000        0.000000         0.000000   
7             0.000000      0.000000        0.000000         0.000000   
8             0.000000      0.000000        0.000000         0.000000   
9             0.000000      0.000000        0.000000         0.000000   
10            0.000000      0.000000        0.000000         0.000000   
11            0.000000      0.000000        0.000000         0.000000   
12            0.000000      0.000000        0.000000         0.000000   
13            0.000000      0.000000        0.000000         0.000000   
14            0.000000      0.000000        0.000000         0.000000   
15            0.000000      0.000000        0.000000         0.000000   
16            0.000000      0.000000        0.000000         0.000000   
17            0.000000      0.000000        0

In [83]:
downtown_grouped.shape

(39, 231)

#### You can  see above that we have 39 neighborhoods with 231 categories. Now lets find the top 5 venues and frequency by neighborhood to review

In [84]:
num_top_venues = 5

for hood in downtown_grouped['Neighborhoods']:
    print("----"+hood+"----")
    temp = downtown_grouped[downtown_grouped['Neighborhoods'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Berczy Park----
                venue  freq
0         Coffee Shop  0.07
1        Cocktail Bar  0.05
2          Restaurant  0.04
3  Italian Restaurant  0.04
4                Café  0.04


----Brockton , Parkdale Village , Exhibition Place----
            venue  freq
0            Café  0.14
1  Breakfast Spot  0.09
2     Coffee Shop  0.09
3       Pet Store  0.05
4       Nightclub  0.05


----Business reply mail Processing CentrE----
                venue  freq
0  Light Rail Station  0.11
1         Yoga Studio  0.06
2                 Spa  0.06
3       Garden Center  0.06
4              Garden  0.06


----CN Tower , King and Spadina , Railway Lands , Harbourfront West , Bathurst Quay , South Niagara , Island airport----
              venue  freq
0   Airport Service  0.17
1    Airport Lounge  0.11
2  Airport Terminal  0.11
3           Airport  0.06
4               Bar  0.06


----Central Bay Street----
                venue  freq
0         Coffee Shop  0.21
1  Italian Restaurant  0.06
2  

#### Lets create a function that will find the most common venues and rank them

In [85]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

#### Now lets run this function an define our variables to create this new data frame with most common venues by neighborhood

In [111]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhoods']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhoods'] = downtown_grouped['Neighborhoods']

for ind in np.arange(downtown_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(downtown_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head(39)

Neighborhoods 1st Most Common Venue  \
0                                         Berczy Park           Coffee Shop   
1      Brockton , Parkdale Village , Exhibition Place                  Café   
2               Business reply mail Processing CentrE    Light Rail Station   
3   CN Tower , King and Spadina , Railway Lands , ...       Airport Service   
4                                  Central Bay Street           Coffee Shop   
5                                            Christie         Grocery Store   
6                                Church and Wellesley      Sushi Restaurant   
7                     Commerce Court , Victoria Hotel           Coffee Shop   
8                                          Davisville        Sandwich Place   
9                                    Davisville North      Department Store   
10                      Dufferin , Dovercourt Village                Bakery   
11            First Canadian Place , Underground city           Coffee Shop   
12                           Forest Hill North & West                  Park   
13                           Garden District, Ryerson        Clothing Store   
14  Harbourfront East , Union Station , Toronto Is...           Coffee Shop   
15                     High Park , The Junction South       Thai Restaurant   
16                    India Bazaar , The Beaches West  Fast Food Restaurant   
17        Kensington Market , Chinatown , Grange Park                  Café   
18                                      Lawrence Park                  Park   
19                          Little Portugal , Trinity                   Bar   
20                       Moore Park , Summerhill East            Playground   
21                                 North Toronto West        Clothing Store   
22                            Parkdale , Roncesvalles             Gift Shop   
23       Queen's Park , Ontario Provincial Government           Coffee Shop   
24                         Regent Park , Harbourfront           Coffee Shop   
25                         Richmond , Adelaide , King           Coffee Shop   
26                                           Rosedale                  Park   
27                                           Roselawn                Garden   
28                                Runnymede , Swansea                  Café   
29                                     St. James Town           Coffee Shop   
30                       St. James Town , Cabbagetown           Coffee Shop   
31                                     Stn A PO Boxes           Coffee Shop   
32                                    Studio District                  Café   
33  Summerhill West , Rathnelly , South Hill , For...           Coffee Shop   
34              The Annex , North Midtown , Yorkville        Sandwich Place   
35                                        The Beaches                 Trail   
36                      The Danforth West , Riverdale      Greek Restaurant   
37          Toronto Dominion Centre , Design Exchange           Coffee Shop   
38                    University of Toronto , Harbord                  Café   

   2nd Most Common Venue          3rd Most Common Venue  \
0           Cocktail Bar             Italian Restaurant   
1         Breakfast Spot                    Coffee Shop   
2            Yoga Studio                            Spa   
3         Airport Lounge               Airport Terminal   
4     Italian Restaurant                           Café   
5                   Café                           Park   
6            Coffee Shop            Japanese Restaurant   
7                   Café                     Restaurant   
8           Dessert Shop                    Coffee Shop   
9                   Park                 Breakfast Spot   
10              Pharmacy                           Park   
11                  Café                          Hotel   
12         Jewelry Store                          Trail   
13           Coffee Shop                           Ca

#### I think I am in luck many of the neighborhoods have cofee shops. Looks like a number of coffee shops ranked as 1st common venue. 12 out of the 39 neighborhoods. We will now set up using kmeans and begin clustering our neighborhoods.

In [112]:
# set number of clusters
kclusters = 5

downtown_grouped_clustering = downtown_grouped.drop('Neighborhoods', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(downtown_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 3], dtype=int32)

#### We will now merge the data from our original data set.

In [113]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

downtown_merged = toronto

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
downtown_merged = downtown_merged.join(neighborhoods_venues_sorted.set_index('Neighborhoods'), on='Neighborhood')

downtown_merged.head() # check the last columns!

PostalCode           Borough                                  Neighborhood  \
0        M5A  Downtown Toronto                    Regent Park , Harbourfront   
1        M7A  Downtown Toronto  Queen's Park , Ontario Provincial Government   
2        M5B  Downtown Toronto                      Garden District, Ryerson   
3        M5C  Downtown Toronto                                St. James Town   
4        M4E      East Toronto                                   The Beaches   

    Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0  43.654260 -79.360636               0           Coffee Shop   
1  43.662301 -79.389494               0           Coffee Shop   
2  43.657162 -79.378937               0        Clothing Store   
3  43.651494 -79.375418               0           Coffee Shop   
4  43.676357 -79.293031               4                 Trail   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                Bakery                   Pub                  Park   
1      Sushi Restaurant                 Diner           Yoga Studio   
2           Coffee Shop                  Café            Restaurant   
3                  Café             Gastropub          Cocktail Bar   
4     Health Food Store                   Pub           Yoga Studio   

  5th Most Common Venue        6th Most Common Venue 7th Most Common Venue  \
0        Breakfast Spot                   Restaurant                  Café   
1              Creperie                     Beer Bar        Sandwich Place   
2   Japanese Restaurant              Bubble Tea Shop        Cosmetics Shop   
3   American Restaurant           Italian Restaurant           Art Gallery   
4         Deli / Bodega  Eastern European Restaurant   Dumpling Restaurant   

       8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0                    Theater    Mexican Restaurant             Shoe Store  
1               Burger Joint         Burrito Place                   Café  
2  Middle Eastern Restaurant                 Diner       Ramen Restaurant  
3         Seafood Restaurant        Clothing Store    Moroccan Restaurant  
4                 Donut Shop      Doner Restaurant                Dog Run

#### Now lets set up our map and plot the clusters.

In [114]:
# create map
map_clusters = folium.Map(location=[latitude_ca, longitude_ca], zoom_start=12)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(downtown_merged['Latitude'], downtown_merged['Longitude'], downtown_merged['Neighborhood'], downtown_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=4,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

### It appears a number of these clusters are alike. I beleive this will need some further analysis. 